File: Rectangle/src/main/java/me/divium/rectangle/Rectangle.java
package me.divium.rectangle;

public class Rectangle {
    private Point firstPoint;
    private Point secondPoint;
    private Color color;

    public Rectangle(Point firstPoint, Point secondPoint, Color color) {
        this.firstPoint = firstPoint;
        this.secondPoint = secondPoint;
        this.color = color;
    }

    public Point getFirstPoint() {
        return firstPoint;
    }

    public Point getSecondPoint() {
        return secondPoint;
    }

    public Color getColor() {
        return color;
    }

    public void setFirstPoint(Point firstPoint) {
        this.firstPoint = firstPoint;
    }

    public void setSecondPoint(Point secondPoint) {
        this.secondPoint = secondPoint;
    }

    public void setColor(Color color) {
        this.color = color;
    }

    @Override
    public String toString() {
        return "Rectangle{" +
                "firstPoint=" + firstPoint +
                ", secondPoint=" + secondPoint +
                ", color=" + color +
                '}';
    }
}

File: RectangleCpp/source/Rectangle.cpp
#include "../include/Rectangle.h"

void Rectangle::validate(Point& firstPoint, Point& secondPoint) {
    if (firstPoint.x == secondPoint.x 
        || firstPoint.y == secondPoint.y)
            throw std::invalid_argument("Rectangle cannot be line or a point");

}

void Rectangle::validate(int width, int height) {
    if (width < 0 || height < 0)
        throw std::invalid_argument("Width or height cannot be < 0");
}

Rectangle::Rectangle(Point firstPoint, Point secondPoint, Color color) {
    validate(firstPoint, secondPoint);
    this->firstPoint = firstPoint;
    this->secondPoint = secondPoint;
    this->color = color;
}

Rectangle::Rectangle(Point firstPoint, int width, int height, Color color) {
    validate(width, height);
    this->firstPoint = firstPoint;
    this->secondPoint = Point(firstPoint.x + width, firstPoint.y + height);
    this->color = color;
}

Point Rectangle::getFirstPoint() {
    return firstPoint;
}

Point Rectangle::getSecondPoint() {
    return secondPoint;
}

Color Rectangle::getColor() {
    return color;
}

void Rectangle::setFirstPoint(Point firstPoint) {
    validate(firstPoint, this->secondPoint);
    this->firstPoint = firstPoint;
}

void Rectangle::setSecondPoint(Point secondPoint) {
    validate(this->firstPoint, secondPoint);
    this->secondPoint = secondPoint;
}

void Rectangle::setColor(Color color) {
    this->color = color;
}

std::ostream &operator<<(std::ostream &out, Rectangle &rectangle) {
    out << rectangle.toString();
    return out;
}

std::string Rectangle::toString() {
    std::stringstream stream;
    stream << "Rectangle{"
           << "firstPoint: " << firstPoint.toString()
           << ", secondPoint: " << secondPoint.toString()
           << ", color: " << color.toString() << "}";
    return stream.str();
}
